src/phoenix/auth.py

```diff
@@ -0,0 +1,66 @@
+import re
+from hashlib import pbkdf2_hmac
+
+from sqlalchemy import select
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from phoenix.db import models
+from phoenix.exceptions import PhoenixException
+
+
+def compute_password_hash(*, password: str, salt: str) -> str:
+    """
+    Salts and hashes a password using PBKDF2, HMAC, and SHA256.
+    """
+    password_bytes = password.encode("utf-8")
+    salt_bytes = salt.encode("utf-8")
+    password_hash_bytes = pbkdf2_hmac("sha256", password_bytes, salt_bytes, NUM_ITERATIONS)
+    password_hash = password_hash_bytes.hex()
+    return password_hash
+
+
+async def validate_login_credentials(
+    *, session: AsyncSession, email: str, password: str, salt: str
+) -> None:
+    """
+    Validates login credentials by computing the password hash and comparing
+    against the password hash stored in the database.
+    """
+    if (
+        user := await session.scalar(select(models.User).where(models.User.email == email))
+    ) is None:
+        raise FailedLoginError
+    assert user.email == email
```

Comment:
intermingling db calls with other validation logic feels like it doesn't need to be here and can be in the route?

Reply:
Not sure what you have in mind. This is just a sanity check that the query on the line above was successful. It shouldn't be necessary, can nix.

Reply:
Broke up this method to avoid computing the password hash inside a transaction.